### PR TITLE
feature (auth/ldap): add LDAP auth method along with associated accounts and managed groups

### PR DESCRIPTION
**Pull request by jimlambrt**
issue: https://github.com/hashicorp/boundary/issues/866

All commits in this PR to main were reviewed as they were merged to the `auth-ldap-llb` branch.

### CODE REVIEW

File: internal/db/schema/migrations/oss/postgres/14/01_wh_user_dimension_oidc.up.sql [hunk 1]
```diff
@@ -39,6 +39,7 @@ begin;
     alter column auth_account_email       type wh_dim_text
   ;
 
+-- Replaced in ???/01_wh_user_dimension_ldap.up.sql
```
Comment (johanbrandhorst) [inline internal/db/schema/migrations/oss/postgres/14/01_wh_user_dimension_oidc.up.sql:42]:
```suggestion
-- Replaced in 64/01_wh_user_dimension_ldap.up.sql
```

File: internal/db/schema/migrations/oss/postgres/9/03_oidc_managed_group_member.up.sql [hunk 2]
```diff
@@ -34,6 +34,7 @@ create trigger default_create_time_column before insert on auth_oidc_managed_gro
 create trigger auth_immutable_managed_oidc_group_member_account before update on auth_oidc_managed_group_member_account
   for each row execute procedure auth_immutable_managed_oidc_group_member_account();
 
+-- Updated in 60/01_ldap.up.sql
```
Comment (johanbrandhorst) [inline internal/db/schema/migrations/oss/postgres/9/03_oidc_managed_group_member.up.sql:37]:
```suggestion
-- Updated in 64/01_ldap.up.sql
```

File: internal/proto/controller/api/resources/authmethods/v1/auth_method.proto [hunk 3]
```diff
@@ -303,3 +308,266 @@ message OidcAuthMethodAuthenticateTokenResponse {
...
+      that: "UseTokenGroups"
+    }
+  ]; // @gotags: `class:"public"`
+
+  // account_attribute_maps are optional attribute maps from custom attributes
+  // to the standard attributes of fullname and email.  These maps are
+  // represented as key=value where the key equals the from_attribute and the
+  // value equals the to_attribute.  For example "preferredName=fullName".  All
+  // attribute names are case insensitive.
+  repeated string account_attribute_maps = 113 [
```
Comment (johanbrandhorst) [inline internal/proto/controller/api/resources/authmethods/v1/auth_method.proto:565]: Tiniest nit: What is this tag number 😛? Looks like it should be 230 to continue the convention.
```suggestion
  repeated string account_attribute_maps = 230 [
```

File: internal/proto/controller/api/resources/managedgroups/v1/managed_group.proto [hunk 4]
```diff
@@ -91,3 +96,16 @@ message OidcManagedGroupAttributes {
     }
   ]; // @gotags: `class:"public"`
 }
+
+// Attributes associated only with ManagedGroups with type "ldap".
+message LdapManagedGroupAttributes {
+  // The list of groups that make up the ManagedGroup
+  repeated string group_names = 100 [
```
Comment (johanbrandhorst) [inline internal/proto/controller/api/resources/managedgroups/v1/managed_group.proto:103]: Nit: the other messages in this file start at a tag number of 10, is there a good reason for this to be 100?

File: internal/proto/controller/api/resources/accounts/v1/account.proto [hunk 5]
```diff
@@ -128,3 +133,39 @@ message OidcAccountAttributes {
   // Output only. userinfo_claims are the marshaled claims from userinfo.
   google.protobuf.Struct userinfo_claims = 130;
 }
+
+// Attributes associated only with Accounts with type "ldap".
+message LdapAccountAttributes {
+  // login_name of the authenticated user.  This is the login_name (or username)
+  // entered by the user when authenticating (typically the uid or cn
+  // attribute).  Account login names must be lower case.
+  string login_name = 100 [
```
Comment (johanbrandhorst) [inline internal/proto/controller/api/resources/accounts/v1/account.proto:142]: Nit: the other messages in this file start at a tag number of 10, is there a good reason for this to start at 100?

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 6]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+  constructed attribute of the user to find the group memberships. This will
+  find all security groups including nested ones.
+
+- `account_attribute_maps` - (optional) If set, the attribute maps from custom
+  attributes to the standard account attributes of fullname and email.  These
+  maps are represented as key=value where the key equals the from_attribute and
+  the value equals the to_attribute.  For example "preferredName=fullName".  All
+  attribute names are case insensitive.
+  
+
```
Comment (johanbrandhorst) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:125]: Probably want to get @Dan-Heath to have a look at these

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

File: internal/db/schema/migrations/oss/postgres/9/03_oidc_managed_group_member.up.sql (hunk 2 above)
Comment (jimlambrt) [reply internal/db/schema/migrations/oss/postgres/9/03_oidc_managed_group_member.up.sql:37]: ty!

File: internal/proto/controller/api/resources/authmethods/v1/auth_method.proto (hunk 3 above)
Comment (jimlambrt) [reply internal/proto/controller/api/resources/authmethods/v1/auth_method.proto:565]: ty!

File: internal/proto/controller/api/resources/managedgroups/v1/managed_group.proto (hunk 4 above)
Comment (jimlambrt) [reply internal/proto/controller/api/resources/managedgroups/v1/managed_group.proto:103]: I prefer to move towards 100 increments and I view this change no different than adding increments of 10 to a proto with no gaps between field numbers.

File: internal/proto/controller/api/resources/accounts/v1/account.proto (hunk 5 above)
Comment (jimlambrt) [reply internal/proto/controller/api/resources/accounts/v1/account.proto:142]: see previous response.

**Review by johanbrandhorst — APPROVED** [review]
I've had about as detailed of a look at this as I can, barring a review on the prose from @Dan-Heath I think this is good to go!

File: website/content/docs/concepts/domain-model/auth-methods.mdx (hunk 6 above)
Comment (jimlambrt) [reply website/content/docs/concepts/domain-model/auth-methods.mdx:125]: Added @Dan-Heath as a reviewer.

File: website/content/docs/concepts/domain-model/accounts.mdx [hunk 7]
```diff
@@ -36,6 +36,33 @@ Password account types have the following additional attributes:
...
+LDAP account types have the following additional attributes:
+
+- `login_name` - (required)
+  Must be unique within the account's [auth method][].
+  Can only contain lower case letters.
+
+- `full_name` - (output only)
+  Maps to the name attribute for the authenticated user and is updated every
+  time the user successfully authenticates. It will be empty until the user's
+  first successful authentication.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/accounts.mdx:50]:
```suggestion
  Maps to the name attribute for the authenticated user, and is updated every
  time the user successfully authenticates. It is empty until the user's
  first successful authentication.
```

File: website/content/docs/concepts/domain-model/accounts.mdx [hunk 8]
```diff
@@ -36,6 +36,33 @@ Password account types have the following additional attributes:
...
+
+- `email` - (output only)
+  Maps to the email address attribute for the authenticated user and is updated
+  every time the user successfully authenticates. It will be empty until the
+  user's first successful authentication.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/accounts.mdx:55]:
```suggestion
  Maps to the email address attribute for the authenticated user, and is updated
  every time the user successfully authenticates. It is empty until the
  user's first successful authentication.
```

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 9]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+- `user_attr` - (optional) If set, defines the attribute on a user's entry 
+  matching the login-name passed when the user authenticates.  Examples: cn, uid  
+
+- `user_filter` - (optional) If set, the go template used to construct an LDAP
+  user search filter. The template can access the following context variables:
+  [UserAttr, Username]. The default user_filter is ({{.UserAttr}}={{.Username}})
+  or (userPrincipalName={{.Username}}@UPNDomain) if the upn-domain parameter is
+  set.
+
+- `enable_groups` - (optional) If true, an authenticated user's groups will be
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:75]:
```suggestion
- `enable_groups` - (optional) If true, an authenticated user's groups are
```

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 10]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+  found during authentication. Defaults to false.
+
+- `group_dn` - (optional) If set, the base DN under which to perform a group
+  search. Example: ou=Groups,dc=example,dc=com
+
+  Note: there is no default, so no base dn will be used for group searches if
+  it's not specified
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:82]:
```suggestion
  Note: There is no default, so no base DN is used for group searches, if
  it's not specified.
```

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 11]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+  (|(memberUid={{.Username}})(member={{.UserDN}})(uniqueMember={{.UserDN}})),
+  which is compatible with several common directory schemas.
+
+- `certificates` - (optional) If set, PEM encoded x509 certificates in ASN.1
+  DER form that can be used as trust anchors when connecting to an LDAP
+  provider. 
+
+- `client_certificate` - (optional) If set, a PEM encoded x509 certificate in
+  ASN.1 DER form to be used as a client certificate. It must be set if the
+  optional client_certificate_key is specified.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:101]:
```suggestion
  ASN.1 DER form to be used as a client certificate. It must be set, if you specify the 
  optional client_certificate_key.
```

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 12]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+
+- `client_certificate_key` - (optional) If set, a PEM encoded certificate key in
+  PKCS #8, ASN.1 DER form. It must be set if the optional client_certificate is 
+  specified.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:105]:
```suggestion
- `client_certificate_key` - (optional) If set, a PEM encoded certificate key in
  PKCS #8, ASN.1 DER form. It must be set, if you specify the optional client_certificate.
```

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 13]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+  
+- `bind_dn` - (optional) If set, the distinguished name of entry to bind when 
+  performing user and group searches. Example:
+  cn=vault,ou=Users,dc=example,dc=com 
+
+- `bind_password` - (optional) If set, the password to use along with bind_dn
+  when performing user search. It must be set if the optional bind_dn is
+  specified.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:113]:
```suggestion
- `bind_password` - (optional) If set, the password to use along with bind_dn
  when performing user search. It must be set, if you specify the optional bind_dn. 
```

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 14]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+  performing user and group searches. Example:
+  cn=vault,ou=Users,dc=example,dc=com 
+
+- `bind_password` - (optional) If set, the password to use along with bind_dn
+  when performing user search. It must be set if the optional bind_dn is
+  specified. 
+
+- `use_token_groups` - (optional) If true, use the Active Directory tokenGroups
+  constructed attribute of the user to find the group memberships. This will
+  find all security groups including nested ones.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:117]:
```suggestion
  constructed attribute of the user to find the group memberships. This 
  finds all security groups, including nested ones.
```

File: website/content/docs/concepts/domain-model/auth-methods.mdx [hunk 15]
```diff
@@ -31,6 +31,98 @@ The password auth method has the following additional attributes:
...
+
+- `account_attribute_maps` - (optional) If set, the attribute maps from custom
+  attributes to the standard account attributes of fullname and email.  These
+  maps are represented as key=value where the key equals the from_attribute and
+  the value equals the to_attribute.  For example "preferredName=fullName".  All
+  attribute names are case insensitive.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/auth-methods.mdx:123]:
```suggestion
- `account_attribute_maps` - (optional) If set, the attribute maps from custom
  attributes to the standard fullname and email account attributes. These
  maps are represented as key=value where the key equals the from_attribute, and
  the value equals the to_attribute.  For example, "preferredName=fullName".  All
  attribute names are case insensitive.
```

File: website/content/docs/concepts/domain-model/managed-groups.mdx [hunk 16]
```diff
@@ -41,6 +41,17 @@ OIDC managed groups have the following additional attributes:
 [filtering concepts]: /boundary/docs/concepts/filtering
 [oidc managed groups filtering]: /boundary/docs/concepts/filtering/oidc-managed-groups
 
+### LDAP Managed Group Information and Attributes
+
+Membership in LDAP managed groups is evaluated when the auth method is used for
+authentication, based on information contained within the LDAP server. Every
+authentication will result in a new evaluation of managed group membership.
```
Comment (Dan-Heath) [inline website/content/docs/concepts/domain-model/managed-groups.mdx:48]:
```suggestion
authentication results in a new evaluation of managed group membership.
```

**Review by Dan-Heath — COMMENTED** [review]
I added some grammatical suggestions to the Markdown docs, but overall thought everything made sense and looks like it's formatted correctly. Please let me know if you have any questions, or if there's anything I can do to help. Thanks!

**Review by Dan-Heath — APPROVED** [review]
LGTM, thank you!

**Comment by jimlambrt** [comment]
Once this PR is merged, we can merge the dependent boundary TF provider PR: https://github.com/hashicorp/terraform-provider-boundary/pull/325

**Pull request merged**